File: attractor/llm/adapter.py
"""Provider adapter protocol."""

from __future__ import annotations

from abc import ABC, abstractmethod
from typing import AsyncIterator

from .models import Request, Response
from .streaming import StreamEvent


class ProviderAdapter(ABC):
    """Interface every LLM provider adapter must implement."""

    @property
    @abstractmethod
    def provider_name(self) -> str: ...

    @abstractmethod
    async def complete(self, request: Request) -> Response: ...

    @abstractmethod
    async def stream(self, request: Request) -> AsyncIterator[StreamEvent]: ...

    async def close(self) -> None:
        """Release resources."""

    async def initialize(self) -> None:
        """Validate configuration on startup."""

    def supports_tool_choice(self, mode: str) -> bool:
        return True

File: internal/llm/adapter.go
package llm

import "context"

// ProviderAdapter is the interface every LLM provider adapter must implement.
type ProviderAdapter interface {
	// ProviderName returns the canonical provider identifier.
	ProviderName() string
	// Complete performs a single non-streaming completion.
	Complete(ctx context.Context, req Request) (Response, error)
	// Stream performs a streaming completion, sending events on the returned
	// channel which is closed when the stream ends.
	Stream(ctx context.Context, req Request) (<-chan StreamEvent, error)
	// Close releases any resources held by the adapter.
	Close() error
}
